cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/engine/SQLEngine.java

```diff
@@ -0,0 +1,107 @@
+  SubmitterLifecycle<BatchContext>, FromKeyValueTransform<StructuredRecord, KEY_IN, VALUE_IN>,
+  ToKeyValueTransform<StructuredRecord, KEY_OUT, VALUE_OUT> {
+  String PLUGIN_TYPE = "sqlengine";
+
+  /**
+   * Creates an Output Format Provided that can be used to push records into a SQL Engine.
+   *
+   * After created, this dataset will be considered "locked" until the output has been committed.
+   *
+   * @param datasetName The name of the dataset to use to store these records.
+   * @param datasetSchema the schema for this dataset.
+   * @return an {@link OutputFormatProvider} instance that can be used to write records to the SQL Engine into the
+   * specified dataset.
+   */
+  OutputFormatProvider getPushProvider(String datasetName, Schema datasetSchema) throws SQLEngineException;
+
+  /**
+   * Creates an InputFormatProvider that can be used to pull records from the specified dataset.
+   * @param datasetName the name of the dataset to pull records from.
+   * @param datasetSchema the schema for this dataset.
+   * @return AN {@link InputFormatProvider} instance that can be used to read records form the specified dataset in the
+   * SQL Engine.
+   */
+  InputFormatProvider getPullProvider(String datasetName, Schema datasetSchema) throws SQLEngineException;
+
+  /**
+   * Check if this dataset exists in the SQL Engine.
+   *
+   * This is a blocking call. if the process to write records into a dataset is ongoing, this method will block until
+   * the process completes. This ensures an accurate result for this operation.
+   *
+   * @param datasetName the dataset name.
+   * @return boolean specifying if this dataset exists in the remote engine.
+   */
+  boolean exists(String datasetName) throws SQLEngineException;
+
+  /**
+   * Check if the supplied Join Definition can be executed in this engine.
+   * @param definition the join definition to validate.
+   * @return boolean specifying if this join operation can be executed in the SQl Engine.
+   */
+  boolean canJoin(JoinDefinition definition);
+
+  /**
+   * Executes the join operation defined by the supplied join Definition and stores the result in the defined dataset.
+   *
+   * The returned {@link Future} can be used to determine the status of this task.
+   *
+   * @param datasetName the name of the dataset to use to store results.
+   * @param definition the join definition, which contains the schema for the resulting dataset.
+   * @return the {@link SQLOperationResult} instance with information about the execution of this task.
+   */
+  SQLOperationResult join(String datasetName, JoinDefinition definition) throws SQLEngineException;
+
+  /**
+   * Deletes all temporary datasets and cleans up all temporary data from the SQL engine.
+   * @param forceStop boolean specifying if all running tasks should be stopped at this time (if any are running).
+   */
+  void cleanup(boolean forceStop);
```

Comment:
I think it's better that cleanup and stopping be associated with each job (through SQLEngineJob.stop() and SQLEngineJob.cleanup() methods).  Making the engine responsible for it means each engine will have to keep track of all running jobs so that it can stop them if forceStop is called.  It would also allow per-job control instead of only allowing stopping all running jobs.

This can be done by making the join() method return a SQLEngineJob rather than the operation result, and adding a getResult() method to the Job class. The SQLEngineJob would then be an interface rather than a concrete class, with the plugin responsible for creating the actual implementation (which would be a thin wrapper around BigQuery's Job class).

Reply:
I have uipdated the cleanup method to support the dataset name as a param.

The caller is then responsible for invoking this method after a run completes.